File: server/watson_assistant/text_to_speech_real_time.js
const record = require('node-record-lpcm16');
const speech = require('@google-cloud/speech');
const client = new speech.SpeechClient();

async function main() {
  console.log("Adjusting for ambient noise... Please wait.");
  console.log("Adjusted. Start speaking.");

  const request = {
    config: {
      encoding: 'LINEAR16',
      sampleRateHertz: 16000,
      languageCode: 'en-US',
    },
    interimResults: false, // If you want interim results, set this to true
  };

  const recognizeStream = client
    .streamingRecognize(request)
    .on('error', (error) => {
      console.error('Error:', error);
    })
    .on('data', (data) => {
      if (data.results[0] && data.results[0].alternatives[0]) {
        console.log(`Recognized: ${data.results[0].alternatives[0].transcript}`);
      } else {
        console.log('Could not understand the audio');
      }
    });

  record
    .start({
      sampleRateHertz: 16000,
      threshold: 0, // Silence threshold
      recordProgram: 'rec', // Try also 'arecord' or 'sox'
      silence: '10.0', // Seconds of silence before ending
    })
    .on('error', console.error)
    .pipe(recognizeStream);

  console.log('Listening...');
}

main().catch(console.error);

File: server/watson_assistant/text_to_speech_real_time.py
import pyaudio
import speech_recognition as sr

def main():
    recognizer = sr.Recognizer()
    mic = sr.Microphone()
    
    with mic as source:
        print("Adjusting for ambient noise... Please wait.")
        recognizer.adjust_for_ambient_noise(source)
        print("Adjusted. Start speaking.")

        while True:
            print("Listening...")
            audio = recognizer.listen(source)

            try:
                text = recognizer.recognize_google(audio)
                print(f"Recognized: {text}")
            except sr.UnknownValueError:
                print("Could not understand the audio")
            except sr.RequestError as e:
                print(f"Could not request results; {e}")

if __name__ == "__main__":
    main()